### ISSUE TITLE
Semantics of deleting a DB on physical data

### ISSUE THREAD

**Issue by e-dard** [influxdata/influxdb_iox#1881]
Currently IOx does not have the ability to delete a database once created. Work is happening now at the API level to provide that operation: https://github.com/influxdata/influxdb_iox/issues/1871

This issue is about describing some useful behaviours that IOx could have when thinking about the side-effects of deleting a database.

Here are some desirable properties from a user perspective:

 - Deleting a database should be a _fast operation_ from the perspective of the API and the user.
 - When a user has deleted a database they should no longer be able to query data from it.
 - If a user accidentally deletes a database it would be nice if they could undo that operation within some time-frame.
 - If a user creates another database with the _same name_, then the new database should *not* contain any of the same state as the previously-deleted database.

Here are some implementation details that could allow us to have some of the properties specified above:

 - When a database is deleted it could be marked in the catalog in some way such that it appears invisible to many operations but is still available to be restored back.
 - The work necessary to remove data in-memory for the database should be asyncronous to the API call ensuring we don't block on drop calls etc.
 - Undoing a deleted database could be a case of re-marking the database in the catalog.
 - If a user creates a subsequent database with the same name then the catalog entry for the old version could be removed entirely. This avoid having to deal with the problem of recovering a database after you created it again.
 - The Object Store would need some way to differentiate between a database's data pre and post-deletion. A marker/generation could do this.

**Comment by domodwyer** [comment]
I like these ideas!

I think you've touched on it, but it might be worth separating out the concerns of deleting a database and _deconfiguring_ a database. The former is obviously permanent (modulo any undelete magic) and the latter is not and would be useful for migrating workloads between machines, running a local IOx with a cut-down set of data, etc.

**Comment by crepererum** [comment]
> When a database is deleted it could be marked in the catalog in some way such that it appears invisible to many operations but is still available to be restored back.

This marker should not be part of the catalog, otherwise to delete a DB with a broken catalog you need to first wipe the catalog which feels a bit cumbersome.

**Comment by alamb** [comment]
If we changed things so that the db name was not hard coded in a fixed location in the the object store path we could then keep all the old state / add new databases without much work at all. We would just have a different object store path

**Comment by pauldix** [comment]
Capturing discussion from this morning, it sounds like this is the path forward. When you create a database its name gets used in the path. E.g. a database named `mydb`:
```
/<writer id>/mydb/...
```
If you delete that database a single empty file gets written:
```
/<writer id>/mydb/DELETED
```

If you then create a database with the same name, it will have a number appended to the end using `-` as a separator:
```
/<writer id>/mydb-1/...
```

If a database name has a `-` in it, it will need to be escaped. Would `%45`, a percent followed by the ascii code for `-`, make sense?

For now, we won't support database renaming. If we want to later, there are ways to do so, but we don't want to offer that up to users for the time being.

**Comment by mkmik** [comment]
> Would %45, a percent followed by the ascii code for -, make sense?

that would be confusing, since we already percent escape all path names in the object store layer, so if we also escape the `-` using percent encoding, we'll end up percent encoding the percent sign

**Comment by crepererum** [comment]
> btw, do we have a defined alphabet for legal characters in a database name?

Partially. Currently the database name checking is implemented here and it only checks for the length and control characters:

https://github.com/influxdata/influxdb_iox/blob/4f5fe62428924da6c43ead3b026b6e9f517fe449/data_types/src/database_name.rs#L55-L77

On top the `object_store` crate does handle characters that are not supported by the cloud provider (I think it uses URL-encoding in most cases). So if the DB name contains a `/`, this will be escaped (we use a path builder instead of simple string concats to ensure that).

**Comment by mkmik** [comment]
I wonder, why do we even have to glue the name and the generation in the same path element, what don't we just do:

`1/my_db/0/....`

When that db is deleted, the next one is created as 


`1/my_db/1/...`

?

**Comment by pauldix** [comment]
Maybe we should just limit db names to alphanumeric and `_`. Then we won't need to escape `-`. Although that's probably not friendly to users that would want to use some broader set of unicode characters (other languages, etc) in their database names.

@mkmik when the database first gets created, it doesn't have the generation in the path. If we use that as a subdir, then we'd need always have the generation in the path. That would certainly solve the escaping issue.

**Comment by mkmik** [comment]
@pauldix yes, let's always have the generation, starting from 0. It's almost no overhead because paths are just strings, directories don't really exist on object stores

**Comment by carols10cents** [comment]
Ok so here's the current implementation plan:

- Changes to database creation: When database `foo_bar` gets created, look for `[server_id]/foo_bar/*`. 
  - If any generation subdirectories exist, check to see if any of them are *missing* the tombstone file (meaning they are active), and if there are, return failure because that database already exists.
  - If all generation subdirectories contain the tombstone file, find the max generation that exists and add 1, then create `[server_id]/foo_bar/[max + 1]/` and return success.
  - If no generation subdirectories exist, create `[server_id]/foo_bar/0` and return success.

- Changes to server startup: Check for tombstone files within generation subdirectories before adding a database to the catalog
  - If there are more than one generation subdirectories missing tombstone files, return an error - this is a bad state. Stop server startup or just refuse to add that database?
  - If there aren't any generation subdirectories missing tombstone files, don't add that database to the catalog at all.

- Add a delete database API that takes a database name. When database `foo_bar` gets deleted,
  - If there is one generation subdirectory that does not contain a tombstone file, add the tombstone file to that generation subdirectory and return success.
  - If there are no generation subdirectories that do not contain a tombstone file, return not found because we can't find any database with that name to delete.
  - If there are multiple generation subdirectories missing tombstone files, this is an inconsistent state-- add tombstone files to all of them to help fix this inconsistent state and return success? Or return failure because we don't know which?
 
- Add a list restoration candidates API that takes a database name (or lists all candidates you have access to?) Name suggestions: `restore_list`, `list_restore_candidates`, `list_deleted_databases`?
  - If there are any generation subdirectories that don't contain tombstone files, return an error because you can't restore a database while there's a current database with that name
  - Return all existing generation subdirectories with helpful metadata like the time they were deleted (which would be the creation time of the tombstone file) or the amount of data in them so the user can pick which one to restore

- Add a restore database API that takes a database name and generation ID gotten from the previous list restore candidates (if this is done in a UI the end user won't necessarily need to see or worry about the generation ID at all, but we need to know which one we're restoring)
  - If there are any generation subdirectories that don't contain tombstone files, return an error because you can't restore a database while there's a current database with that name (prevent race conditions between running list restore candidates and restoring a candidate)
  - Delete the tombstone file in the specified generation subdirectory and return success

- Add a "hard delete" API that takes a database name and generation ID for some external process to call according to some policy for cleaning up old soft-deleted data
  - If the specified database and generation directory doesn't contain a tombstone file, return an error and require soft deletion first to prevent accidents
  - Start a potentially longer running process that cleans up all data in that database's generation directory, deleting the tombstone file last 
  - Provide a way to check on the status of this operation?


Am I missing anything? I feel like I should separate this out into separate tickets.

**Comment by pauldix** [comment]
> If there are more than one generation subdirectories missing tombstone files, return an error - this is a bad state. Stop server startup or just refuse to add that database?

This database should be in an error state on initialization, but server startup should continue

> If there are multiple generation subdirectories missing tombstone files, this is an inconsistent state-- add tombstone files to all of them to help fix this inconsistent state and return success? Or return failure because we don't know which?

Return failure because this database is in a failure state and requires different operator intervention before it can be worked with through the normal API.

I'd make the "hard delete" API just take a database name. More like cleanup tombstoned database. So it would clear out any tombstoned generations that exist for that database name.

I think you've captured it all!

**Comment by alamb** [comment]
This is a really nice (precise and concise) description @carols10cents . It matches my memory of the conversation as well

>  When database foo_bar gets deleted, If there is one generation subdirectory that does not contain a tombstone file, add the tombstone file to that generation subdirectory and return success.

When a "delete database" command is issued, the "active" database (in memory that was successfully loaded) will know what generation number it is, so I think IOx can write the tombstone directly to the active database's catalog directory without having to look at the other generations

**Comment by crepererum** [comment]
Is my understanding correct that the proposal also changes the paths of parquet files and the catalog transaction from the `<server_id>/<db_name>` to `<server_id>/<db_name>/<gen>`? In that case, the paths in `parquet_file` needs to be adjusted as well. We may (or may not) wanna do #2193 before.

**Comment by NGA-TRAN** [comment]
Very nice doc, @carols10cents
One related thing is  can we have an API (or modify the current one) to show only active DBs? Some thing like list_active_databases?

**Comment by carols10cents** [comment]
> One related thing is can we have an API (or modify the current one) to show only active DBs? Some thing like list_active_databases?

Great catch! Adding this-- yes, the management client `list_database` command should only list active databases that have a generation directory without a tombstone file.

**Comment by carols10cents** [comment]
Closing this meta-issue as all parts are either complete or moot :)
